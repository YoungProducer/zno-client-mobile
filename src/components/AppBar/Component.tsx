/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 2 March 2020
 *
 * Header component which contain main routes and another information or actions.
 */

/** External imports */
import React from 'react';
import CssBaseLine from '@material-ui/core/CssBaseline';
import Container from '@material-ui/core/Container';
import AppBar from '@material-ui/core/AppBar';

/** Application's imports */
import { TAppBarProps } from './container';
import NavigationLink from 'components/NavigationLink';

/** Create component */
const Component = (props: TAppBarProps) => {
    const { classes } = props;

    return (
        <React.Fragment>
            <CssBaseLine />
            <AppBar
                position='static'
                className={classes.root}
            >
                <Container
                    maxWidth='lg'
                    className={classes.container}
                >
                    <NavigationLink
                        navLink={{
                            to: '/',
                        }}
                        className={classes.navLink}
                        data-testid='home-navlink'
                    >
                        Головна
                    </NavigationLink>
                    <NavigationLink
                        navLink={{
                            to: '/subject-selection',
                        }}
                        className={classes.navLink}
                        data-testid='tests-navlink'
                    >
                        Тести
                    </NavigationLink>
                </Container>
            </AppBar>
        </React.Fragment>
    );
};

export default Component;
