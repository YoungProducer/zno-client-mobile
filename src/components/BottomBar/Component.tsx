/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 28 March 2020
 *
 * Component which display main navigation of app.
 */

/** External imports */
import React, { useState, useMemo } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import classNames from 'classnames';
import Collapse from '@material-ui/core/Collapse';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import ExpandLessIcon from '@material-ui/icons/ExpandLess';
import HomeIcon from '@material-ui/icons/Home';
import MoreHorizIcon from '@material-ui/icons/MoreHoriz';
import PersonIcon from '@material-ui/icons/Person';
import { makeStyles, createStyles, Theme } from '@material-ui/core';

/** Application's imports */
import { TBottomBarProps } from './container';

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            position: 'absolute',
            minHeight: 70,
            borderRadius: 9,
            background: '#fff',
            filter: 'drop-shadow(0px 3px 3.5px rgba(0,0,0,0.16))',
            bottom: theme.spacing(1),
            left: theme.spacing(1),
            right: theme.spacing(1),
            transform: `translateY(0)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        rootHidden: {
            transform: `translateY(60%)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        innerContainer: {
            width: '100%',
            height: '100%',
        },
        actionsContainer: {
            width: '100%',
            height: '100%',
            minHeight: 70,
            display: 'flex',
            justifyContent: 'space-around',
            alignItems: 'center',
        },
        expandIconButton: {
            zIndex: 10,
            position: 'absolute',
            left: '50%',
            transform: `translate(-50%, -60%)`,
            borderRadius: '50%',
            background: '#fff',
            filter: 'drop-shadow(0px 3px 3.5px rgba(0,0,0,0.05))',
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
            '&:hover': {
                background: '#fff',
            },
        },
        expandIconButtonHidden: {
            transform: `translate(-50%, -35%)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        expandIcon: {
            transform: `rotateX(180deg)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        expandIconHidden: {
            transform: `rotateX(0deg)`,
            transition: theme.transitions.create(['transform'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        actionIconButton: {
            background: '#fff',
            boxShadow: `0px 3px 3.5px rgba(0,0,0,0)`,
            transition: theme.transitions.create(['box-shadow'], {
                duration: 300,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        actionIconButtonActive: {
            boxShadow: `0px 3px 3.5px rgba(0,0,0,0.16)`,
            transition: theme.transitions.create(['box-shadow'], {
                duration: 500,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        actionIcon: {
            color: '#69deac',
            width: 35,
            height: 35,
        },
        expandActions: {
            padding: theme.spacing(1),
            display: 'flex',
            justifyContent: 'space-between',
        },
        button: {
            color: '#fff',
            background: '#656cff',
            borderRadius: 4,
            width: '45%',
        },
    }));

const Component = ({ loggedIn, logout }: TBottomBarProps) => {
    const classes = useStyles({});

    const history = useHistory();
    const location = useLocation();

    const [hidden, setHidden] = useState<boolean>(false);
    const [expand, setExpand] = useState<boolean>(false);

    const toggleHidden = () => {
        if (expand) {
            setExpand(false);
            window.setTimeout(() => {
                setHidden(!hidden);
            }, 300);
        } else {
            setHidden(!hidden);
        }
    };
    const toggleExpand = () => setExpand(!expand);

    const handleSignIn = () => history.push('/auth/signin');
    const handleSignUp = () => history.push('/auth/signup');
    const handleHome = () => history.push('/');
    const handleProfile = () => history.push('/user/profile');

    const activeIcons = useMemo(() => {
        const { pathname } = location;

        return {
            home: pathname === '/' || pathname === '/subject-selection',
            profile: pathname === '/user/profile',
        };
    }, [location]);

    return (
        <div
            className={classNames(classes.root, {
                [classes.rootHidden]: hidden,
            })}
        >
            <IconButton
                className={classNames(classes.expandIconButton, {
                    [classes.expandIconButtonHidden]: hidden,
                })}
                onClick={toggleHidden}
            >
                <ExpandLessIcon className={classNames(classes.expandIcon, {
                    [classes.expandIconHidden]: hidden,
                })}/>
            </IconButton>
            <div className={classes.innerContainer}>
                <div className={classes.actionsContainer}>
                    <IconButton
                        disabled={hidden}
                        className={classNames(classes.actionIconButton, {
                            [classes.actionIconButtonActive]: activeIcons.profile,
                        })}
                        onClick={handleProfile}
                    >
                        <PersonIcon className={classes.actionIcon}/>
                    </IconButton>
                    <IconButton
                        disabled={hidden}
                        className={classNames(classes.actionIconButton, {
                            [classes.actionIconButtonActive]: activeIcons.home,
                        })}
                        onClick={handleHome}
                    >
                        <HomeIcon className={classes.actionIcon}/>
                    </IconButton>
                    <IconButton
                        disabled={hidden}
                        className={classNames(classes.actionIconButton, {
                            [classes.actionIconButtonActive]: expand,
                        })}
                        onClick={toggleExpand}
                    >
                        <MoreHorizIcon className={classes.actionIcon}/>
                    </IconButton>
                </div>
                <Collapse in={expand} unmountOnExit>
                    <div className={classes.expandActions}>
                        { loggedIn && (
                            <Button
                                className={classes.button}
                                variant='contained'
                                disableElevation
                                onClick={logout}
                            >
                                Вийти
                            </Button>
                        )}
                        { !loggedIn && (
                            <>
                                <Button
                                    className={classes.button}
                                    variant='contained'
                                    disableElevation
                                    onClick={handleSignIn}
                                >
                                    Увійти
                                </Button>
                                <Button
                                    className={classes.button}
                                    variant='contained'
                                    disableElevation
                                    onClick={handleSignUp}
                                >
                                    Зареєструватися
                                </Button>
                            </>
                        )}
                    </div>
                </Collapse>
            </div>
        </div>
    );
};

export default Component;
