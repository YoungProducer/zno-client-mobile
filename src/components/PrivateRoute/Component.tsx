/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 1 March 2020
 *
 * Permission route which automaticaly redirect from protected pages if user is not logged in to SignIn page.
 */

/** External imports */
import React from 'react';
import { Redirect, Route } from 'react-router-dom';

/** Application's imports */
import { TPrivateRoute } from './container';

/** Create component */
const Component = ({ isLoggedIn, children, ...rest }: TPrivateRoute) => (
    <Route
        {...rest}
        render={({ location }) => {
            const pathname = location
                ? `/auth/signin?next=${location.pathname}${location.search}`
                : '/auth/signin';

            return isLoggedIn
                ? children
                : <Redirect
                    to={{
                        pathname: '/auth/signin',
                        state: { from: location },
                    }}
                />;
        }
        }
    />
);

export default Component;
