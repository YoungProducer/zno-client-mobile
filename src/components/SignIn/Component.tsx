/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 28 February 2020
 *
 * Component which allows user to make sign in
 */

/** External imports */
import React, { useCallback, useState, useEffect } from 'react';
import classNames from 'classnames';
import Paper from '@material-ui/core/Paper';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import EmailIcon from '@material-ui/icons/Email';
import LockIcon from '@material-ui/icons/Lock';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import { useHistory } from 'react-router-dom';

/** Application's imports */
import { TSignInProps } from './container';
import NavigationLink from 'components/NavigationLink';
import Input from 'components/custom/Input';
import Checkbox from 'components/custom/Checkbox';
import Logo from 'img/logo';

/** Define classes as hook */
const useStyles = makeStyles((theme: Theme) => createStyles({
    container: {
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        width: '100%',
        boxSizing: 'border-box',
        padding: theme.spacing(1),
    },
    root: {
        borderRadius: 4,
        width: '100%',
        filter: `drop-shadow(2.828px 2.828px 3.5px rgba(227,207,207,0.9))`,
        backgroundColor: '#ffffff',
        textAlign: 'center',
    },
    innerContainer: {
        borderRadius: 4,
        backgroundColor: '#f4f5f7',
        // padding: theme.spacing(4),
    },
    logoContainer: {
        paddingTop: theme.spacing(1),
        paddingBottom: theme.spacing(1),
    },
    textField: {
        width: 230,
    },
    fields: {
        paddingBottom: theme.spacing(3),
    },
    actions: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
    },
    block: {
        width: '100%',
        margin: 0,
        '& .MuiGrid-item': {
            width: '100%',
        },
    },
    buttonBlock: {
        width: '100%',
        marginTop: theme.spacing(2),
    },
    button: {
        filter: `drop-shadow(0px 3px 2.5px rgba(0,0,0,0.16))`,
    },
    link: {
        color: theme.palette.primary.main,
    },
    icon: {
        color: '#555',
    },
    inputRoot: {
        width: '100%',
        height: 60,
        background: `#fff`,
        filter: `drop-shadow(0px 3px 2.5px rgba(0,0,0,0.16))`,
        borderRadius: 4,
        paddingLeft: theme.spacing(2),
        paddingRight: theme.spacing(2),
    },
    checkBox: {
        width: '100%',
    },
}));

/**
 * Create hook to make component stateless.
 * Compute all values from text fields.
 * Returns fields data and button onClick property.
 */
const useSignInElements = (props: TSignInProps) => {
    const {
        isLoggedIn,
        errorFields,
        fieldsMessages,
        fetchSignIn,
        setSignInErrorFieldsToDefault,
        setSignInFieldsMessagesToDefault,
    } = props;

    const history = useHistory();

    useEffect(() => {
        if (isLoggedIn) {
            const next = history.location.state
                ? `${(history.location.state as any).from.pathname}${(history.location.state as any).from.search}`
                : undefined;

            history.push(next || '/');
        }
    }, [isLoggedIn]);

    const clearFields = () => {
        setSignInErrorFieldsToDefault();
        setSignInFieldsMessagesToDefault();
    };

    /** User email */
    const [email, setEmail] = useState<string>('');
    const handleChangeEmail =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setEmail(event.target.value);
            if (errorFields.email) {
                clearFields();
            }
        }, [errorFields.email]);

    /** User password */
    const [password, setPassword] = useState<string>('');
    const handleChangePassword =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setPassword(event.target.value);
            if (errorFields.password) {
                clearFields();
            }
        }, [errorFields.password]);

    /** Remember checkbox */
    const [remember, setRemember] = useState<boolean>(false);
    const handleChangeRemember =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setRemember(event.target.checked);
        }, []);

    /** SignUp button handle */
    const handleFetchSignIn = useCallback(() => fetchSignIn({
        email,
        password,
        remember,
    }), [email, password, remember]);

    return {
        emailField: {
            onChange: handleChangeEmail,
            value: email,
            error: errorFields.email,
            helperText: fieldsMessages.email,
        },
        passwordField: {
            onChange: handleChangePassword,
            value: password,
            error: errorFields.password,
            helperText: fieldsMessages.password,
        },
        rememberCheckbox: {
            onChange: handleChangeRemember,
            value: remember,
        },
        signInButton: {
            onClick: handleFetchSignIn,
        },
    };
};

/** Create component */
const Component = (props: TSignInProps) => {
    /** Create styles */
    const classes = useStyles({});

    /** Get fields data from hook */
    const { emailField, passwordField, rememberCheckbox, signInButton } = useSignInElements(props);

    return (
        <div
            className={classNames('auth-background', classes.container)}
        >
            <Paper
                className={classes.root}
                elevation={0}
                square
            >
                <div className={classes.logoContainer}>
                    <Logo />
                </div>
                <div className={classes.innerContainer}>
                    <Grid
                        container
                        direction='column'
                        alignItems='center'
                        justify='flex-start'
                        item
                        spacing={3}
                        className={classes.block}
                    >
                        <Grid item>
                            <Input
                                color='primary'
                                type='email'
                                placeholder='Емеїл'
                                startAdornment={<EmailIcon className={classes.icon}/>}
                                data-testid='signin-email-input'
                                rootClassName={classes.inputRoot}
                                {...emailField}
                            />
                        </Grid>
                        <Grid item>
                            <Input
                                color='primary'
                                type='password'
                                placeholder='Пароль'
                                startAdornment={<LockIcon className={classes.icon}/>}
                                data-testid='signin-password-input'
                                rootClassName={classes.inputRoot}
                                {...passwordField}
                            />
                        </Grid>
                        <Grid item style={{
                            width: 'inherit',
                        }}>
                            <Checkbox
                                {...rememberCheckbox}
                                label={`Запам'ятати мене`}
                                formControl={{
                                    className: classes.checkBox,
                                }}
                            />
                        </Grid>
                        <Grid item>
                            <Button
                                color='primary'
                                variant='contained'
                                disableElevation
                                className={classes.button}
                                data-testid='signin-button'
                                { ...signInButton }
                            >
                                Увійти
                            </Button>
                        </Grid>
                        <Grid item>
                            <Typography align='center'>
                                Не зареєстровані?
                                <NavigationLink
                                    navLink={{
                                        to: '/auth/signup',
                                    }}
                                    className={classes.link}
                                >
                                    Зареєструватися
                                </NavigationLink>
                            </Typography>
                        </Grid>
                    </Grid>
                </div>
            </Paper>
        </div>
    );
};

export default Component;
