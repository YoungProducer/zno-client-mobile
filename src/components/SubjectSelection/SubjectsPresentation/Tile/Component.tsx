/**
 * Created by: Oleksandr Bezrukov
 * Creation datel: 24 March 2020
 *
 * Component which render task tiles with correct colors.
 */

/** External imports */
import React, { useState, useCallback, useEffect, useMemo } from 'react';
import { useHistory } from 'react-router-dom';
import classNames from 'classnames';
import Typography from '@material-ui/core/Typography';
import ButtonBase from '@material-ui/core/ButtonBase';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import { yellow, red } from '@material-ui/core/colors';

/** Application's imports */
import { AdditionalAnswerPropertiesContext } from 'context/TestSuiteContext';
import { TTileProps } from './container';

export const tileWidth = 251;
export const activeTileWidth = 251;
/** Distance between tiles */
export const tilesSpacing = 16;

const useTileStyles = makeStyles((theme: Theme) =>
    createStyles({
        tile: {
            minWidth: tileWidth,
            width: tileWidth,
            height: 230,
            borderRadius: 17,
            overflow: 'overlay',
            boxShadow: `0px 0.5px 3.5px rgba(0, 0, 0, 0.16)`,
            background: '#edeff3',
            fontSize: '1.25rem',
            color: '#b19898',
            // filter: `drop-shadow(0px 3px 3.5px rgba(0,0,0,0.16))`,
            display: 'block',
            transition: theme.transitions.create(['height', 'width', 'background', 'color'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
            marginRight: tilesSpacing,
            '&:last-child': {
                marginRight: 0,
            },
        },
        tileActive: {
            // height: 230,
            width: activeTileWidth,
            minWidth: activeTileWidth,
            background: '#fff',
            filter: `drop-shadow(0px 3px 3.5px rgba(0,0,0,0.16))`,
            transition: theme.transitions.create(['height', 'width', 'background', 'color'], {
                duration: 200,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
        tileHide: {
            visibility: 'hidden',
        },
        img: {
            width: '100%',
            height: 145,
            objectFit: 'contain',
            verticalAlign: 'top',
        },
        desc: {
            background: '#fff',
            height: 85,
            width: '100%',
            padding: theme.spacing(1),
        },
        label: {
            color: '#777b7e',
        },
    }));

const Tile = ({
    taskIndex,
    active,
    callback,
    hide,
    selected,
    gived,
    right,
    finished,
    image,
}: TTileProps) => {
    const classes = useTileStyles({});

    return (
        <div
            className={classes.tile}
            onClick={callback}
        >
            <img src={image} className={classes.img}/>
            <div className={classes.desc}>
                <Typography variant='h5' className={classes.label}>
                    Математика
                </Typography>
            </div>
        </div>
    );
};

export default Tile;
