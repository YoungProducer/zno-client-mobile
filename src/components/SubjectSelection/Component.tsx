/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 1 March 2020
 *
 * Component to select subjects.
 */

/** External imports */
import React, { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import classNames from 'classnames';
import AppBar from '@material-ui/core/AppBar';
import Container from '@material-ui/core/Container';
import Button from '@material-ui/core/Button';
import InputBase from '@material-ui/core/InputBase';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import SearchIcon from '@material-ui/icons/Search';
import { makeStyles, Theme, createStyles } from '@material-ui/core';

/** Application's imports */
import { TSubjectSelectionProps } from './container';
import SubjectTile from './SubjectTile';
import SubjectPresentation from './SubjectsPresentation';
import Logo from 'img/logo';

const useStyles = makeStyles((theme: Theme) => createStyles({
    root: {
        width: '100%',
        height: '100vh',
    },
    container: {
        display: 'flex',
        marginTop: theme.spacing(10),
    },
    searchBlock: {
        width: '50%',
    },
    popularBlock: {
        background: '#5c498c',
        height: 85,
        width: 'max-content',
        position: 'absolute',
        // bottom: '1%',
        bottom: '1%',
        right: 0,
        left: 0,
        margin: 'auto',
        borderRadius: 7,
        paddingLeft: theme.spacing(3),
        paddingRight: theme.spacing(3),
    },
    popularBlockInner: {
        height: 'inherit',
        width: '100%',
    },
    searchInput: {
        width: '100%',
        height: 67,
        background: '#4d8fcb',
        borderRadius: 34,
        paddingLeft: theme.spacing(2),
        paddingRight: 6,
        color: '#fff',
        fontSize: '1.3rem',
        marginTop: theme.spacing(8),
    },
    iconButton: {
        width: 55,
        height: 55,
        background: '#fff',
        borderRadius: '50%',
        '&:hover': {
            background: '#f1f1f1',
        },
    },
    icon: {
        color: '#5955cf',
        width: 30,
        height: 30,
    },
    typography: {
        color: '#fff',
    },
    actionsBlock: {
        display: 'flex',
        alignItems: 'center',
    },
    appBar: {
        background: 'none',
        paddingTop: theme.spacing(3),
        paddingBottom: theme.spacing(3),
    },
    appBarContainer: {
        display: 'flex',
        justifyContent: 'space-between',
    },
    appBarActions: {

    },
    signUpButton: {
        borderRadius: 25,
        background: '#fff',
        color: theme.palette.primary.main,
        paddingLeft: theme.spacing(2),
        paddingRight: theme.spacing(2),
        '&:hover': {
            background: '#f1f1f1',
        },
        minWidth: 100,
    },
    signInButton: {
        minWidth: 100,
        color: '#fff',
        '&:hover': {
            background: 'none',
        },
    },
}));

/** Create component */
const Component = (props: TSubjectSelectionProps) => {
    const classes = useStyles({});

    const history = useHistory();

    const {
        subjectsList,
        loggedIn,
        fetchSubjectsNames,
        fetchLogout,
    } = props;

    const [seachValue, setSearchValue] = useState<string>('');
    const handleSearchChange = (event: React.ChangeEvent<HTMLInputElement>) =>
        setSearchValue(event.target.value);

    useEffect(() => {
        fetchSubjectsNames();
    }, []);

    return (
        <>
            <div className={classNames(classes.root, 'subject-selection-background')}>
                <AppBar elevation={0} className={classes.appBar} position='relative'>
                    <Container maxWidth='lg' className={classes.appBarContainer}>
                        <Logo />
                        <div className={classes.appBarContainer}>
                            { !loggedIn && (
                                <>
                                    <Button
                                        className={classes.signInButton}
                                        onClick={() => history.push('/auth/signin')}
                                    >
                                        Увійти
                                    </Button>
                                    <Button
                                        variant='text'
                                        className={classes.signUpButton}
                                        onClick={() => history.push('/auth/signup')}
                                    >
                                        Зареєструватися
                                    </Button>
                                </>
                            )}
                            { loggedIn && (
                                <Button
                                    variant='text'
                                    className={classes.signUpButton}
                                    onClick={fetchLogout}
                                >
                                    Вийти
                                </Button>
                            )}
                        </div>
                    </Container>
                </AppBar>
                <Container maxWidth='lg' className={classes.container}>
                    <div className={classes.actionsBlock}>
                        <div className={classes.searchBlock}>
                            <Typography
                                variant='h6'
                                className={classes.typography}
                            >
                                Для того щоб розпочати тест почніть вводити назву предемету у пошуку, після цього ви одразу побачите цей предмет
                                у правому блоці, просто натисніть на нього і ви будете направлені на сторінку тесту.
                            </Typography>
                            <InputBase
                                className={classes.searchInput}
                                placeholder='Знайти предмет'
                                value={seachValue}
                                onChange={handleSearchChange}
                                endAdornment={
                                    <IconButton
                                        size='small'
                                        className={classes.iconButton}
                                    >
                                        <SearchIcon
                                            className={classes.icon}
                                        />
                                    </IconButton>
                                }
                            />
                        </div>
                        <SubjectPresentation
                            subjectsList={subjectsList}
                            searchValue={seachValue}
                        />
                    </div>
                </Container>
            </div>
            {/* <div className={classes.popularBlock}>
                <SubjectTile
                    subject={subjectsList[0]}
                />
                <SubjectTile
                    subject={subjectsList[1]}
                />
            </div> */}
        </>
    );
};

export default Component;
