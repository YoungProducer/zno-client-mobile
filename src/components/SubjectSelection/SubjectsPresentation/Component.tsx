/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 23 March 2020
 *
 * Component which allows to select which task user want to complete.
 */

/** External imports */
import React, { useState, useCallback, useEffect, useMemo } from 'react';
import { useHistory } from 'react-router-dom';
import classNames from 'classnames';
import Box from '@material-ui/core/Box';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

/** Application's imports */
import Tile, {
    tileWidth,
    activeTileWidth,
    tilesSpacing,
} from './Tile';
import math from 'images/math.jpg';

const useComponentStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            overflow: 'hidden',
            position: 'relative',
            height: 300,
            width: '100%',
            cursor: 'pointer',
            marginLeft: theme.spacing(-1),
        },
        innerContainer: {
            display: 'flex',
            alignItems: 'flex-start',
            position: 'absolute',
            height: '100%',
            width: '100%',
            paddingLeft: theme.spacing(1),
        },
        animateContainer: {
            transition: theme.transitions.create('left', {
                duration: 500,
                easing: theme.transitions.easing.easeInOut,
            }),
        },
    }));

const useCarouselFields = (props: ITaskSelectionProps) => {
    /** Destruct props */
    const {
        tasksAmount,
        activeTask,
        setTaskIndex,
    } = props;

    const history = useHistory();

    const handleTileClick = (id: string) => history.push(`/subject-configuration/${id}`);

    /** Offset of the inner container */
    const [offsetX, setOffestX] = useState<number>(0);
    const [prevStartX, setPrevStartX] = useState<number>(0);
    /** Toggles when mouse down or up */
    const [mouseDown, toggleMouseDown] = useState<boolean>(false);
    /** Related to animate class in inner container */
    const [animate, toggleAnimate] = useState<boolean>(false);

    const rootContainerRef = React.createRef<HTMLDivElement>();

    const [buttonsAmount, setButtonsAmount] = useState<number>(0);

    const minOffset = 0;
    const maxOffset = tasksAmount * (tileWidth + tilesSpacing) - tilesSpacing + (activeTileWidth - tileWidth);

    /**
     * If offset is out of range then
     * set offset in extreme positions.
     */
    const handleSetOffset = (offset: number) => {
        if (offset > minOffset) {
            setOffestX(minOffset);
        } else if (offset < -maxOffset + rootContainerRef.current.clientWidth) {
            setOffestX(-maxOffset + rootContainerRef.current.clientWidth);
        } else {
            setOffestX(offset);
        }
    };

    /**
     * Calculate diapason in which tiles must be displayed
     * to optimize carousel if tile index is out of range
     * to this tile will be added class with 'visibility: hidden'.
     */
    const getFirstTileIndex = useCallback(() => {
        const index = Math.floor(offsetX / (tileWidth + tilesSpacing)) * -1;
        return 0 > index ? 0 : index - 2;
    }, [offsetX]);

    const getLastTileIndex = useCallback(() => {
        const index = buttonsAmount + Math.floor(offsetX / (tileWidth + tilesSpacing)) * -1;
        return buttonsAmount > index ? buttonsAmount : index + 2;
    }, [buttonsAmount, offsetX]);

    /**
     * Handle mouse move to drag the carousel
     * offset divided by 2 to make carousel slower.
     */
    const handleOnTouchStart = (event: React.TouchEvent<HTMLDivElement>) => {
        setPrevStartX(event.touches[0].clientX);
    };

    const handleOnTouchMove = (event: React.TouchEvent<HTMLDivElement>) => {
        event.persist();
        const currentTouchX = event.touches[0].clientX;

        const mouseOffsetX = (prevStartX - currentTouchX) / -1;
        const newOffsetX = offsetX + mouseOffsetX;

        setPrevStartX(currentTouchX);

        handleSetOffset(newOffsetX);
    };

    /**
     * Toggle animate to 'false' to remove
     * animation class in innerContainer
     * to prevent bugs.
     */
    const handleOnMouseDown = () => {
        toggleMouseDown(true);
        toggleAnimate(false);
    };

    const handleOnMouseUp = () => toggleMouseDown(false);

    /**
     * If new active tile has border index
     * then toggle animate to 'true' to add
     * class in innerContainer which allows
     * to animate 'left' property and then
     * calculate and add offset to move
     * tiles to left or to right.
     */
    useEffect(() => {
        if (!mouseDown && activeTask !== 0) {
            /**
             * Difference between last tile and new active el.
             * 2 - lastTileIndex offset.
             */
            const lastDiff = getLastTileIndex() - 2 - activeTask;
            /**
             * Difference between first tile and new active el.
             * 2 - firstTileIndex offset.
             */
            const firstDiff = getFirstTileIndex() + 2 - activeTask;

            if (lastDiff < 3 && lastDiff >= 0) {
                toggleAnimate(true);
                handleSetOffset(Math.floor(offsetX - (buttonsAmount - 2) * (tileWidth + tilesSpacing)));
            }

            if (firstDiff < 3 && firstDiff >= 0) {
                toggleAnimate(true);
                handleSetOffset(Math.floor(offsetX + (buttonsAmount - 2) * (tileWidth + tilesSpacing)));
            }
        }
        if (activeTask <= getFirstTileIndex()) {
            const firstDiff = getFirstTileIndex() + 2;

            if (firstDiff !== 0) {
                toggleAnimate(true);
                handleSetOffset(Math.floor(offsetX + (getLastTileIndex() - 2) * (tileWidth + tilesSpacing)));
            }
        }
    }, [activeTask]);

    /**
     * Calculate buttons amount per innerContainer width.
     */
    useEffect(() => {
        const containerWidth = rootContainerRef.current.clientWidth;

        const amount = ((containerWidth + tilesSpacing) / (tileWidth + tilesSpacing));

        setButtonsAmount(Math.floor(amount));
    }, [rootContainerRef]);

    const tiles = useMemo(() =>
        [{ image: math, id: '5e69049424aa9a00072762eb' }].map((_, index) => (
            <Tile
                key={index}
                image={_.image}
                taskIndex={index}
                active={index === activeTask}
                hide={!(index >= getFirstTileIndex() && index <= getLastTileIndex())}
                callback={() => {
                    handleTileClick(_.id);
                }}
            />
        )), [
            activeTask,
            tasksAmount,
            getFirstTileIndex,
            getLastTileIndex,
        ]);

    return {
        handleOnTouchMove,
        handleOnMouseDown,
        handleOnMouseUp,
        handleOnTouchStart,
        tiles,
        animate,
        offsetX,
        rootContainerRef,
        buttonsAmount,
    };
};

export interface ITaskSelectionProps {
    tasksAmount: number;
    setTaskIndex: (index: number) => void;
    activeTask: number;
}

const Component = (props: ITaskSelectionProps) => {
    const classes = useComponentStyles({});

    const {
        offsetX,
        animate,
        tiles,
        rootContainerRef,
        handleOnTouchMove,
        handleOnTouchStart,
    } = useCarouselFields(props);

    return (
        <div
            className={classes.root}
            // onTouchStart={handleOnTouchStart}
            // onTouchMove={handleOnTouchMove}
            ref={rootContainerRef}
        >
            <Box
                className={classNames(classes.innerContainer, {
                    [classes.animateContainer]: animate,
                })}
                left={offsetX}
            >
                {tiles}
            </Box>
        </div>
    );
};

export default Component;
