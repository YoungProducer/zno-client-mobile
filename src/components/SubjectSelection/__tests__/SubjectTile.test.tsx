/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 1 March 2020
 *
 * Create test suites for SubjectTile component.
 * @jest-environment jsdom
 */

/** External imports */
import React from 'react';
import { shallow } from 'enzyme';

/** Application's imports */
import Component, { ISubjectTileProps } from '../SubjectTile';

/** Create mock function for useHistory hook */
const push = jest.fn();

/** Mock react-router-dom module */
jest.mock('react-router-dom', () => ({
    useHistory: jest.fn(() => ({
        push,
    })),
}));

describe('SubjectTile component', () => {
    /** Define required props */
    const requiredProps: ISubjectTileProps = {
        subject: {
            id: 'foo',
            name: 'bar',
        },
    };

    test('Is match snapshot', () => {
        /** Render via enzyme */
        const root = shallow(<Component {...requiredProps} />);

        /** Assert component matches snapshot */
        expect(root).toMatchSnapshot();
    });

    test('Button onClick handler', () => {
        /** Render via enzyme */
        const tree = shallow(<Component {...requiredProps}/>);

        /** Simulate */
        tree.find(`[data-testid='subject-tile-button']`).simulate('click');

        expect(push).toHaveBeenCalledWith('/subject-configuration/foo');
    });
});
