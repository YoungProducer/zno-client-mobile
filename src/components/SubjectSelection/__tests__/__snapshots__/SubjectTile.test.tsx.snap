// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubjectTile component Is match snapshot 1`] = `
<WithStyles(ForwardRef(Button))
  className="makeStyles-button-1"
  data-testid="subject-tile-button"
  onClick={[Function]}
  variant="text"
>
  bar
</WithStyles(ForwardRef(Button))>
`;
